File: core/src/main/java/com/draniksoft/ome/editor/systems/pos/PhysicsSys.java
package com.draniksoft.ome.editor.systems.pos;

import com.artemis.Aspect;
import com.artemis.BaseEntitySystem;
import com.artemis.ComponentMapper;
import com.artemis.EntitySubscription;
import com.artemis.annotations.Wire;
import com.artemis.utils.IntBag;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.physics.box2d.*;
import com.draniksoft.ome.editor.components.pos.PhysC;
import com.draniksoft.ome.editor.components.pos.PosSizeC;
import com.draniksoft.ome.editor.components.tps.MObjectC;
import com.draniksoft.ome.utils.PUtils;
import com.draniksoft.ome.utils.struct.Pair;

public class PhysicsSys extends BaseEntitySystem {

    @Wire
    World pw;

    ComponentMapper<PhysC> pM;
    ComponentMapper<PosSizeC> p2M;

    int passA = 10;
    int passS = passA;


    public PhysicsSys() {
        super(Aspect.all(PhysC.class));
    }

    @Override
    protected void initialize() {

        getSubscription().addSubscriptionListener(new EntitySubscription.SubscriptionListener() {
            @Override
            public void inserted(IntBag entities) {

            }

            @Override
            public void removed(IntBag es) {

                for (int i = 0; i < es.size(); i++) {
                    pw.destroyBody(pM.get(es.get(i)).b);
                }

            }
        });

    }

    @Override
    protected void processSystem() {

        if (passS > 0) {
            passS--;
            return;
        }

        pw.step(Gdx.graphics.getDeltaTime(), 2, 2);

        passS = passA;
    }

    // returns physics pos in map coords
    public Pair<Float, Float> getPhysPos(int _e) {

        if (!pM.has(_e)) return null;

        Body b = pM.get(_e).b;

        return Pair.createPair(b.getPosition().x * PUtils.PPM, b.getPosition().y * PUtils.PPM);

    }

    public void setCentricPos(float x, float y, int _e) {
	  if (pM.has(_e)) {

		pM.get(_e).b.setTransform(x / PUtils.PPM, y / PUtils.PPM, pM.get(_e).b.getAngle());

	  }
    }

    // positions in the center
    public void setSyncCentricPos(float x, float y, int _e) {

	  setCentricPos(x, y, _e);

	  if (p2M.has(_e)) {
		PosSizeC c = p2M.get(_e);
		c.x = (int) x - c.w / 2;
		c.y = (int) y - c.h / 2;
	  }


    }

    public void createBodyFromSPos(int x, int y, int _e) {

	  PosSizeC pc = p2M.get(_e);

	  PhysC c = pM.create(_e);

	  BodyDef bd = new BodyDef();
	  bd.type = BodyDef.BodyType.KinematicBody;
	  bd.position.set((x) / PUtils.PPM, (y) / PUtils.PPM);
	  Body b = world.getInjector().getRegistered(com.badlogic.gdx.physics.box2d.World.class).createBody(bd);
	  b.setUserData(_e);
	  PolygonShape pg = new PolygonShape();
	  pg.setAsBox(pc.w / 2 / PUtils.PPM, pc.h / 2 / PUtils.PPM);
	  FixtureDef fd = new FixtureDef();
	  fd.shape = pg;
	  b.createFixture(fd);
	  pg.dispose();
	  c.b = b;

	  setCentricPos(pc.x + pc.w / 2f, pc.y + pc.h / 2f, _e);


    }

    public void saveMOSyncPos(float x, float y, int _e) {

	  setSyncCentricPos(x, y, _e);

        if (!world.getMapper(MObjectC.class).has(_e)) return;

        MObjectC c = world.getMapper(MObjectC.class).get(_e);
	  c.x = (int) x - c.w / 2;
	  c.y = (int) y - c.h / 2;

    }

    public void syncPos(int _e) {

	  setSyncCentricPos(getPhysPos(_e).getElement0(), getPhysPos(_e).getElement1(), _e);

    }
}

File: core/src/main/java/com/draniksoft/ome/editor/systems/pos/PhysicsSys.kt
package com.draniksoft.ome.editor.systems.pos

import com.artemis.Aspect
import com.artemis.BaseEntitySystem
import com.artemis.ComponentMapper
import com.artemis.EntitySubscription
import com.artemis.annotations.Wire
import com.artemis.utils.IntBag
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.FixtureDef
import com.badlogic.gdx.physics.box2d.PolygonShape
import com.badlogic.gdx.physics.box2d.World
import com.draniksoft.ome.editor.components.pos.PhysC
import com.draniksoft.ome.editor.components.pos.PosSizeC
import com.draniksoft.ome.editor.components.tps.MObjectC
import com.draniksoft.ome.utils.PUtils
import com.draniksoft.ome.utils.struct.Pair


class PhysicsSys : BaseEntitySystem(Aspect.all(PhysC::class.java)) {

    @Wire
    internal var pw: World? = null

    internal var pM: ComponentMapper<PhysC>? = null
    internal var p2M: ComponentMapper<PosSizeC>? = null

    internal var passA = 10
    internal var passS = passA

    override fun initialize() {

        getSubscription().addSubscriptionListener(object : EntitySubscription.SubscriptionListener {
            override fun inserted(entities: IntBag) {

            }

            override fun removed(es: IntBag) {

                for (i in 0..es.size() - 1) {
                    pw!!.destroyBody(pM!!.get(es.get(i)).b)
                }

            }
        })

    }

    override fun processSystem() {

        if (passS > 0) {
            passS--
            return
        }

        pw!!.step(Gdx.graphics.deltaTime, 2, 2)

        passS = passA
    }

    // returns physics pos in map coords
    fun getPhysPos(_e: Int): Pair<Float, Float>? {

        if (!pM!!.has(_e)) return null

        val b = pM!!.get(_e).b

        return Pair.createPair(b.position.x * PUtils.PPM, b.position.y * PUtils.PPM)

    }

    fun setCentricPos(x: Float, y: Float, _e: Int) {
        if (pM!!.has(_e)) {

            pM!!.get(_e).b.setTransform(x / PUtils.PPM, y / PUtils.PPM, pM!!.get(_e).b.angle)

        }
    }

    // positions in the center
    fun setSyncCentricPos(x: Float, y: Float, _e: Int) {

        setCentricPos(x, y, _e)

        if (p2M!!.has(_e)) {
            val c = p2M!!.get(_e)
            c.x = x.toInt() - c.w / 2
            c.y = y.toInt() - c.h / 2
        }


    }

    fun createBodyFromSPos(x: Int, y: Int, _e: Int) {

        val pc = p2M!!.get(_e)

        val c = pM!!.create(_e)

        val bd = BodyDef()
        bd.type = BodyDef.BodyType.KinematicBody
        bd.position.set(x / PUtils.PPM, y / PUtils.PPM)
        val b = world.injector.getRegistered(com.badlogic.gdx.physics.box2d.World::class.java).createBody(bd)
        b.userData = _e
        val pg = PolygonShape()
        pg.setAsBox(pc.w.toFloat() / 2f / PUtils.PPM, pc.h.toFloat() / 2f / PUtils.PPM)
        val fd = FixtureDef()
        fd.shape = pg
        b.createFixture(fd)
        pg.dispose()
        c.b = b

        setCentricPos(pc.x + pc.w / 2f, pc.y + pc.h / 2f, _e)


    }

    fun saveMOSyncPos(x: Float, y: Float, _e: Int) {

        setSyncCentricPos(x, y, _e)

        if (!world.getMapper(MObjectC::class.java).has(_e)) return

        val c = world.getMapper(MObjectC::class.java).get(_e)
        c.x = x.toInt() - c.w / 2
        c.y = y.toInt() - c.h / 2

    }

    fun syncPos(_e: Int) {

        setSyncCentricPos(getPhysPos(_e)!!.element0, getPhysPos(_e)!!.element1, _e)

    }
}
